Subject: [PATCH] Fix batterySendBroadcast not synchronized issue.

Symptom: system server crash.
Root Cause: batterySendBroadcast not sychrnized with AMS object,
caused system crash in broadcastIntentLocked()
when 2 threads access mParallelBroadcasts.get(i).intent but get null
in BroadcastQueue.replaceParallelBroadcastLocked().
Solution: add sychronization object.

diff --git a/services/core/java/com/android/server/am/ActivityManagerService.java b/services/core/java/com/android/server/am/ActivityManagerService.java
index c589da4..c6f8b9a 100644
--- a/services/core/java/com/android/server/am/ActivityManagerService.java
+++ b/services/core/java/com/android/server/am/ActivityManagerService.java
@@ -3113,9 +3113,11 @@
 
     @Override
     public void batterySendBroadcast(Intent intent) {
-        broadcastIntentLocked(null, null, intent, null, null, 0, null, null, null,
-                AppOpsManager.OP_NONE, null, false, false,
-                -1, Process.SYSTEM_UID, UserHandle.USER_ALL);
+        synchronized (this) {
+            broadcastIntentLocked(null, null, intent, null, null, 0, null, null, null,
+                    AppOpsManager.OP_NONE, null, false, false,
+                    -1, Process.SYSTEM_UID, UserHandle.USER_ALL);
+        }
     }
 
     /**
